lib/arm/arm_lifter.ml

```diff
@@ -227,7 +227,8 @@ let lift_move ~encoding mem ops (insn : move_insn) : stmt list =
 
   | `MOVTi16, [|`Reg dest; _; src; cond; _wflag|] ->
     let dest = Env.of_reg dest in
-    [Bil.move dest Bil.(var dest lor exp_of_op src lsl int32 16)] |>
+    let src16 = Bil.(cast unsigned 32 (cast low 16 (var dest))) in
+    [Bil.move dest Bil.(src16 lor exp_of_op src lsl int32 16)] |>
```

Comment:
It is better not to abuse casts for bit arithmetic. Casts should be used only to change the size of the word. It is much easier to analyze bit-wise operations (especially for SMT solvers) than casts, so my suggestion is to encode it as,

```
movt r0, #0x1c
{
  R0 := R0 & 0xFFFF | 0x1C0000
}
```
i.e.,

```suggestion
    let dest = Env.of_reg dest and src = exp_of_op src in
    Bil.[dest := var dest land int32 0xFFFF lor src lsl int32 16] |>
```

Reply:
👍